cleverhans/model_zoo/deep_k_nearest_neighbors/dknn.py

```diff
@@ -0,0 +1,429 @@
+    """
+    Implements the DkNN algorithm. See https://arxiv.org/abs/1803.04765 for more details.
+
+    :param neighbors: number of neighbors to find per layer.
+    :param layers: a list of layer names to include in the DkNN.
+    :param get_activations: a callable that takes a np array and a layer name and returns its activations on the data.
+    :param train_data: a np array of training data.
+    :param train_labels: a np vector of training labels.
+    :param nb_classes: the number of classes in the task.
+    :param scope: a TF scope that was used to create the underlying model.
+    :param nb_tables: number of tables used by FALCONN to perform locality-sensitive hashing.
+    """
+    super(DkNNModel, self).__init__(nb_classes=nb_classes, scope=scope)
+    self.neighbors = neighbors
+    self.nb_tables = nb_tables
+    self.layers = layers
+    self.get_activations = get_activations
+    self.nb_cali = -1
+    self.calibrated = False
+
+    # Compute training data activations
+    self.nb_train = train_labels.shape[0]
+    assert self.nb_train == train_data.shape[0]
+    self.train_activations = get_activations(train_data)
+    self.train_labels = train_labels
+
+    # Build locality-sensitive hashing tables for training representations
+    self.train_activations_lsh = copy.copy(self.train_activations)
+    self.init_lsh()
+
+  def init_lsh(self):
+    """
+    Initializes locality-sensitive hashing with FALCONN to find nearest neighbors in training data.
+    """
+    self.query_objects = {
+    }  # contains the object that can be queried to find nearest neighbors at each layer.
+    # mean of training data representation per layer (that needs to be substracted before LSH).
+    self.centers = {}
+    for layer in self.layers:
+      assert self.nb_tables >= self.neighbors
+
+      # Normalize all the lenghts, since we care about the cosine similarity.
+      self.train_activations_lsh[layer] /= np.linalg.norm(
+          self.train_activations_lsh[layer], axis=1).reshape(-1, 1)
+
+      # Center the dataset and the queries: this improves the performance of LSH quite a bit.
+      center = np.mean(self.train_activations_lsh[layer], axis=0)
+      self.train_activations_lsh[layer] -= center
+      self.centers[layer] = center
+
+      # LSH parameters
+      params_cp = falconn.LSHConstructionParameters()
+      params_cp.dimension = len(self.train_activations_lsh[layer][1])
+      params_cp.lsh_family = falconn.LSHFamily.CrossPolytope
+      params_cp.distance_function = falconn.DistanceFunction.EuclideanSquared
+      params_cp.l = self.nb_tables
+      params_cp.num_rotations = 2  # for dense set it to 1; for sparse data set it to 2
+      params_cp.seed = 5721840
+      # we want to use all the available threads to set up
+      params_cp.num_setup_threads = 0
+      params_cp.storage_hash_table = falconn.StorageHashTable.BitPackedFlatHashTable
+
+      # we build 18-bit hashes so that each table has
```

Comment:
Maybe worth parameterizing this if it's not meant to be MNIST/CIFAR specific

Reply:
Yes, that's a good idea